Add unit test for watch function in resourcesVfs

jobe-m opened the pull request:
The added "watch" unit test should pass since the watch call of
resourcesVfs is inside a launch lambda. Currently it fails because of a
bug in LocalVfsJvm.kt (most probably...).

This is for reproducing issue korlibs/korge#727

jobe-m committed:
Add unit test for watch function in resourcesVfs

The added "watch" unit test should pass since the watch call of
resourcesVfs is inside a launch lambda. Currently it fails because of a
bug in LocalVfsJvm.kt (most probably...).

soywiz commented:
That test seems to prevent gradle to finish

jobe-m commented:
Yes, exactly. This test is for reproducing the issue mentioned above.
Interestingly the "watch" function is called inside "launch" and the test is still suspending the current thread.
I thing that the watch functionality is somehow handlwed wrongly. When I have time I will try to figure this behaviour.

jobe-m committed:
Changed watcher function to use Dispatcher IO

The implementation of the watch functionality in the Jvm implementation
was changed to use Dispatcher IO instead of the executeIO function. That
was inspired by:

https://github.com/vishna/watchservice-ktx/blob/master/src/main/kotlin/dev/vishna/watchservice/watchservice.kt

Now it looks like that the watch function is working. But it is not
suspending any more the main thread. That means no launch is needed in
the test case.

Maybe later it would be an idea to use also coroutines.channels as API
to the application.

soywiz commented on korio/src/jvmMain/kotlin/com/soywiz/korio/file/std/LocalVfsJvm.kt:
```diff
@@ -402,22 +402,16 @@ private class LocalVfsJvm : LocalVfsV2() {
+                val key = watcher.take()
 
 				for (e in key.pollEvents()) {
 					val kind = e.kind()
```
Could you wrap `when (kind) { ... }` with `withContext(coroutineContext) { when (kind) { ... } }` that way handler will be executed in the right thread where the function was called, in korio and korge this is important since it is mostly single threaded and mostly not thread-safe

jobe-m committed:
Fix review finding

Wrapping the callback of the handler into coroutineContext so that the
handler will be executed in the right thread where the function was
called. This is important since korio and korge are mostly implemented
single threaded and without coroutineContext the code would not be
thread-safe.

soywiz commented:
LGTM. Thanks! 

![korio_watching](https://user-images.githubusercontent.com/570848/117668075-dec36680-b1a5-11eb-89c4-5b1165be3932.jpg)

jobe-m commented:
Thanks for merging :)
